core/src/either.rs

```diff
@@ -194,32 +57,54 @@ where
         self: Pin<&mut Self>,
         cx: &mut Context<'_>,
     ) -> Poll<Result<Self::Substream, Self::Error>> {
-        match self.project() {
-            EitherOutputProj::First(inner) => inner
+        match as_pin_mut(self) {
+            future::Either::Left(inner) => inner
                 .poll_outbound(cx)
-                .map_ok(EitherOutput::First)
+                .map_ok(future::Either::Left)
                 .map_err(Either::Left),
-            EitherOutputProj::Second(inner) => inner
+            future::Either::Right(inner) => inner
                 .poll_outbound(cx)
-                .map_ok(EitherOutput::Second)
+                .map_ok(future::Either::Right)
                 .map_err(Either::Right),
         }
     }
 
     fn poll_close(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
-        match self.project() {
-            EitherOutputProj::First(inner) => inner.poll_close(cx).map_err(Either::Left),
-            EitherOutputProj::Second(inner) => inner.poll_close(cx).map_err(Either::Right),
+        match as_pin_mut(self) {
+            future::Either::Left(inner) => inner.poll_close(cx).map_err(Either::Left),
+            future::Either::Right(inner) => inner.poll_close(cx).map_err(Either::Right),
         }
     }
 
     fn poll(
         self: Pin<&mut Self>,
         cx: &mut Context<'_>,
     ) -> Poll<Result<StreamMuxerEvent, Self::Error>> {
-        match self.project() {
-            EitherOutputProj::First(inner) => inner.poll(cx).map_err(Either::Left),
-            EitherOutputProj::Second(inner) => inner.poll(cx).map_err(Either::Right),
+        match as_pin_mut(self) {
+            future::Either::Left(inner) => inner.poll(cx).map_err(Either::Left),
+            future::Either::Right(inner) => inner.poll(cx).map_err(Either::Right),
+        }
+    }
+}
+
+/// Convert `Pin<&mut Either<A, B>>` to `Either<Pin<&mut A>, Pin<&mut B>>`,
+/// pinned projections of the inner variants.
+///
+/// Local function until <https://github.com/rust-lang/futures-rs/pull/2691> is merged.
+fn as_pin_mut<A, B>(
```

Comment:
Should we ask for [project](https://docs.rs/futures-util/0.3.25/src/futures_util/future/either.rs.html#36) to be public? As it seems the same purpose and would allow us to not have unsafe code here.

Reply:
> Should we ask for [project](https://docs.rs/futures-util/0.3.25/src/futures_util/future/either.rs.html#36) to be public? As it seems the same purpose and would allow us to not have unsafe code here.

See https://github.com/rust-lang/futures-rs/issues/2629#issuecomment-1196338344 :)

This is the `future::Either` type, we can't use `either::Either` here because it doesn't implement `AsyncRead` etc

Reply:
yeah I saw after, thanks Thomas :)  
It it would be nice for `either::Either` to implement the `futures` traits as you asked on https://github.com/bluss/either/issues/79. But it seems you also submitted https://github.com/rust-lang/futures-rs/pull/2691 which will allow us to remove the `unsafe` code right? Ideally what would be really nice was for a single `Either` type :grinning: which will be possible when https://github.com/rust-lang/futures-rs/pull/2691 is merged right? Nonetheless thanks for your work simplifying this Thomas!

Reply:
> yeah I saw after, thanks Thomas :)  
> It it would be nice for `either::Either` to implement the `futures` traits as you asked on https://github.com/bluss/either/issues/79. But it seems you also submitted https://github.com/rust-lang/futures-rs/pull/2691 which will allow us to remove the `unsafe` code right?

Correct!

> Ideally what would be really nice was for a single `Either` type :grinning: which will be possible when https://github.com/rust-lang/futures-rs/pull/2691 is merged right?

Not quite unfortunately, we also use `Either` in places where it needs to implement `std::error::Error` for example and `futures::Either` doesn't AFAIK. I agree that it would be nice but at least after these patches, we don't have a problem with there being two different `Either` types.

> Nonetheless thanks for your work simplifying this Thomas!

You are welcome ☺️

Reply:
wdyt of then leaving a `TODO` on the unsafe code to be removed when a `futures` gets released with https://github.com/rust-lang/futures-rs/pull/2691 ?

Reply:
> wdyt of then leaving a `TODO` on the unsafe code to be removed when a `futures` gets released with https://github.com/rust-lang/futures-rs/pull/2691 ? 

Do you mean like line 93? 🙈